packages/react-router-dom/__tests__/data-browser-router-test.tsx

```diff
@@ -3000,6 +3000,280 @@ function testDomRouter(
+        it("useSubmit", async () => {
+          let router = createTestRouter(
+            createRoutesFromElements(
+              <Route
+                id="parent"
+                path="/parent"
+                element={<Parent />}
+                action={({ request }) => "PARENT ACTION: " + request.url}
+              >
+                <Route
+                  id="index"
+                  index
+                  element={<Index />}
+                  action={({ request }) => "INDEX ACTION: " + request.url}
+                />
+              </Route>
+            ),
+            {
+              window: getWindow("/parent?index&index=keep"),
+            }
+          );
+          let { container } = render(<RouterProvider router={router} />);
+
+          function Parent() {
+            let actionData = useActionData();
+            let submit = useSubmit();
+            return (
+              <>
+                <p id="parent">{actionData}</p>
+                <button onClick={() => submit({}, { method: "post" })}>
+                  Submit from parent
+                </button>
+                <Outlet />
+              </>
+            );
+          }
+
+          function Index() {
+            let actionData = useActionData();
+            let submit = useSubmit();
+            return (
+              <>
+                <p id="index">{actionData}</p>
+                <button onClick={() => submit({}, { method: "post" })}>
+                  Submit from index
+                </button>
+              </>
+            );
+          }
+
+          fireEvent.click(screen.getByText("Submit from parent"));
+          await tick();
+          await waitFor(() => screen.getByText(new RegExp("PARENT ACTION")));
+          expect(getHtml(container.querySelector("#parent")!)).toContain(
+            "PARENT ACTION: http://localhost/parent?index=keep"
+          );
+
+          fireEvent.click(screen.getByText("Submit from index"));
+          await tick();
+          await waitFor(() => screen.getByText(new RegExp("INDEX ACTION")));
+          expect(getHtml(container.querySelector("#index")!)).toContain(
+            "INDEX ACTION: http://localhost/parent?index&index=keep"
+          );
+        });
+
+        it("Form", async () => {
+          let router = createTestRouter(
+            createRoutesFromElements(
+              <Route
+                id="parent"
+                path="/parent"
+                element={<Parent />}
+                action={({ request }) => "PARENT ACTION: " + request.url}
+              >
+                <Route
+                  id="index"
+                  index
+                  element={<Index />}
+                  action={({ request }) => "INDEX ACTION: " + request.url}
+                />
+              </Route>
+            ),
+            {
+              window: getWindow("/parent?index&index=keep"),
```

Comment:
`Form`/`fetcher.Form` did not have this issue since `useFormAction` handled it - but the logic in `useFormAction`to remove the `?index` param was naive and removed all index params, not preserving user-specified ones.  This PR also fixes that issue